backend/kube/kube.go

```diff
@@ -0,0 +1,201 @@
+// Package kube is a backend implementation that uses the Tinkerbell CRDs to get DHCP data.
+package kube
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+
+	"github.com/tinkerbell/dhcp/data"
+	"github.com/tinkerbell/tink/pkg/apis/core/v1alpha1"
+	"github.com/tinkerbell/tink/pkg/controllers"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+	"inet.af/netaddr"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+)
+
+const tracerName = "github.com/tinkerbell/dhcp"
+
+// Backend is a backend implementation that uses the Tinkerbell CRDs to get DHCP data.
+type Backend struct {
+	cluster cluster.Cluster
+}
+
+// NewBackend returns a controller-runtime cluster.Cluster with the Tinkerbell runtime
+// scheme registered, and indexers for:
+// * Hardware by MAC address
+//
+// Callers must instantiate the client-side cache by calling StartCache() before use.
+func NewBackend(conf *rest.Config, opts ...cluster.Option) (*Backend, error) {
+	rs := runtime.NewScheme()
+
+	if err := scheme.AddToScheme(rs); err != nil {
+		return nil, err
+	}
+
+	if err := v1alpha1.AddToScheme(rs); err != nil {
+		return nil, err
+	}
+
+	opts = append([]cluster.Option{func(o *cluster.Options) { o.Scheme = rs }}, opts...)
+	o := []cluster.Option{func(o *cluster.Options) { o.Scheme = rs }}
+	o = append(o, opts...)
+	c, err := cluster.New(conf, o...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new cluster config: %w", err)
+	}
+
+	if err := c.GetFieldIndexer().IndexField(context.Background(), &v1alpha1.Hardware{}, ".spec.interfaces.dhcp.mac", controllers.HardwareMacIndexFunc); err != nil {
+		return nil, fmt.Errorf("failed to setup indexer: %w", err)
+	}
+
+	return &Backend{cluster: c}, nil
+}
+
+// StartCache starts the client-side cache.
+func (b *Backend) StartCache(ctx context.Context) error {
```

Comment:
We might want to leave this as "Start" because the thing we're starting is actually an informer that deposits a specific type of object into a cache. Leaving it intentionally vague has the benefit of not assuming what's happening under the hood, instead the "Backend" needs starting.